### PR TITLE
fix: update preprocessing pipelines

### PR DESCRIPTION
**Pull request by ktagowski**

### CODE REVIEW

**Review by asawczyn — DISMISSED** [review]

File: embeddings/data/dataset.py [hunk 1]
```diff
@@ -23,14 +21,7 @@ def __repr__(self) -> str:
...
 
-class LocalDataset(Dataset[T_path]):
-    def __init__(self, dataset: T_path, **load_dataset_kwargs: Any):
-        super().__init__()
-        self.dataset = dataset
-        self.load_dataset_kwargs = load_dataset_kwargs
-
-
-class HuggingFaceDataset(Dataset[str]):
+class LoadableDataset(Dataset[str]):
```
Comment (djaniak) [inline embeddings/data/dataset.py:24]: I'd consider renaming `LoadableDataset` to some other name since probably all of our datasets will be loadable (is there any use of unloadable dataset?)

File: embeddings/data/data_loader.py [hunk 2]
```diff
@@ -34,29 +33,30 @@ def load(self, dataset: Dataset[str]) -> datasets.DatasetDict:
 
 class HuggingFaceLocalDataLoader(DataLoader[str, datasets.DatasetDict]):
     def load(self, dataset: Dataset[str]) -> datasets.DatasetDict:
-        if isinstance(dataset, HuggingFaceDataset):
+        if isinstance(dataset, LoadableDataset):
             result = datasets.load_from_disk(str(dataset.dataset))
```
Comment (djaniak) [inline embeddings/data/data_loader.py:37]:
```suggestion
            result = datasets.load_from_disk(dataset.dataset)
```
dataset is a `str` attribute

File: embeddings/data/data_loader.py [hunk 3]
```diff
@@ -74,9 +74,13 @@ def load(self, dataset: Dataset[T_path]) -> Corpus:
...
-def get_hf_dataloader(
-    dataset: HuggingFaceDataset,
-) -> Union[HuggingFaceDataLoader, HuggingFaceLocalDataLoader]:
+HF_DATALOADERS = Union[HuggingFaceDataLoader, HuggingFaceLocalDataLoader]
+FLAIR_DATALOADERS = Union[
+    HuggingFaceDataLoader, PickleFlairCorpusDataLoader, ConllFlairCorpusDataLoader
+]
+
+
+def get_hf_dataloader(dataset: LoadableDataset) -> HF_DATALOADERS:
```
Comment (djaniak) [inline embeddings/data/data_loader.py:83]: how much work would it take to create a `AutoDataLoader` class (such as the ones for embedding)? do you think it would be a good idea?

File: embeddings/data/dataset.py (hunk 1 above)
Comment (ktagowski) [reply embeddings/data/dataset.py:24]: I agree with that. I named it based on variables. Any suggestion how to improve it?

File: embeddings/data/data_loader.py (hunk 3 above)
Comment (ktagowski) [reply embeddings/data/data_loader.py:83]: I think that's not so easy, because we would need to add functionality that detects whether given dataset is hf datasets or flair datasets. Without that it would be not so `auto`

**Comment by djaniak** [comment]
preprocessing pipelines are in two seperate files: hf_preprocessing_pipeline and preprocessing_pipeline; the latter contains also flair preprocessing pipelines
i'd either put them in 3 separate files or in 1 common

File: embeddings/pipeline/preprocessing_pipeline.py [hunk 4]
```diff
@@ -52,120 +58,126 @@ def run(self) -> TransformationResult:
...
+    seed: int = 441
+    load_dataset_kwargs: Optional[Dict[str, Any]] = None
 
-class FlairTextPairClassificationPreprocessingPipeline(
-    PreprocessingPipeline[str, datasets.DatasetDict, Corpus]
-):
-    def __init__(
+    def __post_init__(self) -> None:
+        self._setup()
+        dataset, data_loader, transformation = self._setup_transformations()
```
Comment (djaniak) [inline embeddings/pipeline/preprocessing_pipeline.py:93]: I dont like the idea that _setup_transformations() returns objects such as dataset, data_loader
It should only setup the transformation so in my understanding it shouldn't even return transformation but set self.transformation attribute

File: embeddings/pipeline/preprocessing_pipeline.py [hunk 5]
```diff
@@ -52,120 +58,126 @@ def run(self) -> TransformationResult:
...
+        super(FlairPreprocessingPipeline, self).__init__(dataset, data_loader, transformation)
+
+    @abc.abstractmethod
+    def _setup(self) -> None:
+        pass
+
+    def _setup_transformations(
```
Comment (djaniak) [inline embeddings/pipeline/preprocessing_pipeline.py:100]: I'd completely refactor this method. It is too long and does other things beside setting up transformations (setting up dataloader and dataset)

Comment (djaniak) [reply embeddings/pipeline/preprocessing_pipeline.py:100]: I'd split these functionalities into a _setup for the base class and setup_basic_transfromation and setup_auxiliary_transformation or something like that were in each class we could define own basic transformation and get rid of this long if else statement

File: embeddings/pipeline/preprocessing_pipeline.py [hunk 6]
```diff
@@ -52,120 +58,126 @@ def run(self) -> TransformationResult:
...
-        transformation = transformation.persisting(FlairConllPersister(path=persist_path))
-        super().__init__(dataset, data_loader, transformation)
+        transformation = transformation.persisting(self.persister)
+
+        return dataset, data_loader, transformation
+
+
+class FlairTextClassificationPreprocessingPipeline(FlairPreprocessingPipeline):
+    def _setup(self) -> None:
+        self.persister = FlairPicklePersister(self.persist_path)
```
Comment (djaniak) [inline embeddings/pipeline/preprocessing_pipeline.py:170]: If we define `FlairPreprocessingPipeline` as a dataclass we can do the same here and set the persister as a class attribute. Then we are refactor the _setup and _setup_transformation methods as stated in the previous comment.

File: tests/test_lightning_classification_pipeline.py [hunk 7]
```diff
@@ -16,18 +17,29 @@ def pipeline_kwargs() -> Dict[str, Any]:
...
+        downsample_splits=(0.01, 0.01, 0.05),
+        seed=441,
+    )
+    pipeline.run()
+
+    return {
+        "dataset_name_or_path": path.name,
+        "input_column_name": ["text"],
+        "target_column_name": "target",
+    }, path
```
Comment (djaniak) [inline tests/test_lightning_classification_pipeline.py:42]: do we need the path to be returned if later in the pipeline we use `dataset_kwargs[0]`?

File: tests/test_lightning_sequence_labeling_pipeline.py [hunk 8]
```diff
@@ -6,17 +6,30 @@
...
+    )
+    pipeline.run()
+
     return {
-        "dataset_name_or_path": "clarin-pl/kpwr-ner",
+        "dataset_name_or_path": path.name,
         "input_column_name": "tokens",
         "target_column_name": "ner",
-    }
+    }, path
```
Comment (djaniak) [inline tests/test_lightning_sequence_labeling_pipeline.py:32]: do we need the path to be returned if later in the pipeline we use `dataset_kwargs[0]`?

File: embeddings/data/dataset.py (hunk 1 above)
Comment (djaniak) [reply embeddings/data/dataset.py:24]: maybe rename this one to Dataset and Dataset to BaseDataset?

File: embeddings/data/data_loader.py [hunk 9]
```diff
@@ -34,29 +33,30 @@ def load(self, dataset: Dataset[str]) -> datasets.DatasetDict:
...
 class HuggingFaceLocalDataLoader(DataLoader[str, datasets.DatasetDict]):
     def load(self, dataset: Dataset[str]) -> datasets.DatasetDict:
-        if isinstance(dataset, HuggingFaceDataset):
-            result = datasets.load_from_disk(str(dataset.dataset))
+        if isinstance(dataset, LoadableDataset):
+            result = datasets.load_from_disk(dataset.dataset)
             assert isinstance(result, datasets.DatasetDict)
             return result
         else:
             raise ValueError("This DataLoader should be used with HuggingFaceDataset only.")
```
Comment (AdrianSzymczak) [inline embeddings/data/data_loader.py:41]: is this error to be adjusted and handled?

File: embeddings/data/datamodule.py [hunk 10]
```diff
@@ -103,13 +103,12 @@ def prepare_data(self) -> None:
 
     def setup(self, stage: Optional[str] = None) -> None:
         self.dataset = self.load_dataset()
-        self.downsample_dataset()
```
Comment (AdrianSzymczak) [inline embeddings/data/datamodule.py:106]: maybe not now but consider comments and logs to communicate to users what's happening

File: embeddings/data/dataset.py (hunk 1 above)
Comment (AdrianSzymczak) [reply embeddings/data/dataset.py:24]: consider Base Dataset and Expandable / Extensible Dataset? Base is fine but just Dataset seems a little to general

File: embeddings/pipeline/hf_preprocessing_pipeline.py [hunk 11]
```diff
@@ -27,16 +26,17 @@ def __init__(
         dataset_name: str,
         persist_path: str,
         sample_missing_splits: Optional[Tuple[Optional[float], Optional[float]]] = None,
+        downsample_splits: Optional[
+            Tuple[Optional[float], Optional[float], Optional[float]]
```
Comment (AdrianSzymczak) [inline embeddings/pipeline/hf_preprocessing_pipeline.py:30]: Consider adding new file with just common types definitions to import as now you repeat quite long definitions in couple of places like here and in embeddings/pipeline/preprocessing_pipeline.py etc

File: embeddings/pipeline/preprocessing_pipeline.py [hunk 12]
```diff
@@ -52,120 +58,134 @@ def run(self) -> TransformationResult:
         return result
 
 
-class FlairTextClassificationPreprocessingPipeline(
-    PreprocessingPipeline[str, datasets.DatasetDict, Corpus]
+# Mypy currently properly don't handle dataclasses with abstract methods  https://github.com/python/mypy/issues/5374
```
Comment (AdrianSzymczak) [inline embeddings/pipeline/preprocessing_pipeline.py:61]: Consider adding todo and followup task

File: embeddings/pipeline/preprocessing_pipeline.py [hunk 13]
```diff
@@ -52,120 +58,134 @@ def run(self) -> TransformationResult:
...
+    )
+
+    dataset_name_or_path: str
+    persist_path: str
+    input_column_name: Union[str, Tuple[str, str]]
+    target_column_name: str
+    datasets_path: Path = DATASET_PATH
+    sample_missing_splits: Optional[Tuple[Optional[float], Optional[float]]] = None
+    downsample_splits: Optional[Tuple[Optional[float], Optional[float], Optional[float]]] = None
+    downsample_splits_stratification: bool = True
```
Comment (AdrianSzymczak) [inline embeddings/pipeline/preprocessing_pipeline.py:78]: how about prefixing boolean variables with should_ or if_ etc?

File: embeddings/transformation/hf_transformation/downsample_transformation.py [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+
+class DownsampleHuggingFaceTransformation(
+    Transformation[datasets.DatasetDict, datasets.DatasetDict]
+):
+    def __init__(
+        self,
+        downsample_train: Optional[float] = None,
+        downsample_dev: Optional[float] = None,
+        downsample_test: Optional[float] = None,
+        seed: int = 441,
```
Comment (AdrianSzymczak) [inline embeddings/transformation/hf_transformation/downsample_transformation.py:16]: setting default seed to 441 in many places shouldn't be unified?

File: tests/hps/test_hps_classification.py [hunk 15]
```diff
@@ -191,6 +230,9 @@ def test_hparams_best_params_files_compatibility(
         "classifier_dropout",
     }
     for k in common_keys:
+        if k == "dataset_name_or_path":
+            continue
+
         if k in metadata:
             if isinstance(metadata[k], Enum):  # type: ignore
                 enum_hparam = getattr(type(metadata[k]), hparams[k])  # type: ignore
```
Comment (AdrianSzymczak) [inline tests/hps/test_hps_classification.py:238]: at some that would be nice if these asserts would produce human readable info what to do if they don't pass

File: tests/test_callback_best_epoch.py [hunk 16]
```diff
@@ -16,18 +17,29 @@ def pipeline_kwargs() -> Dict[str, Any]:
...
+        sample_missing_splits=None,
+        ignore_test_subset=False,
+        downsample_splits=(0.01, 0.01, 0.05),
+        seed=441,
+    )
+    pipeline.run()
+
+    return {
+        "dataset_name_or_path": path.name,
+        "input_column_name": ["text"],
```
Comment (AdrianSzymczak) [inline tests/test_callback_best_epoch.py:40]: I would expect single "text" because of name suffix and I wouldn't expect list -> consider adding s as suffix so that it indicates plural names

File: tests/test_callback_best_epoch.py [hunk 17]
```diff
@@ -71,7 +78,7 @@ def lightning_classification_pipeline(
         LightningClassificationPipeline(
             output_path=result_path.name,
             **pipeline_kwargs,
-            **dataset_kwargs,
+            **dataset_kwargs[0],
```
Comment (AdrianSzymczak) [inline tests/test_callback_best_epoch.py:81]: I'd consider adding comments for future us in couple of months

**Review by AdrianSzymczak — CHANGES_REQUESTED** [review]

File: tests/test_lightning_classification_pipeline.py (hunk 7 above)
Comment (ktagowski) [reply tests/test_lightning_classification_pipeline.py:42]: Otherwise the temporary path is destroyed. Passing `path.name` is not sufficient to keep object

File: tests/test_lightning_sequence_labeling_pipeline.py (hunk 8 above)
Comment (ktagowski) [reply tests/test_lightning_sequence_labeling_pipeline.py:32]: Same reason as previous

File: embeddings/data/data_loader.py (hunk 9 above)
Comment (ktagowski) [reply embeddings/data/data_loader.py]: Yes. I removed it as it's no longer needed. In current version of dataloaders we do not differentiate that

File: embeddings/data/datamodule.py (hunk 10 above)
Comment (ktagowski) [reply embeddings/data/datamodule.py:106]: It's general setup function which is in `pytorch_lightning` package. Here we just load dataset and we converts it to torch format. Although we can add logging here, what is your opinion?

File: embeddings/pipeline/hf_preprocessing_pipeline.py (hunk 11 above)
Comment (ktagowski) [reply embeddings/pipeline/hf_preprocessing_pipeline.py]: Done, added it to `__init__` file in `pipelines` module

File: embeddings/pipeline/preprocessing_pipeline.py (hunk 12 above)
Comment (ktagowski) [reply embeddings/pipeline/preprocessing_pipeline.py]: Good idea, but it's outside of scope of this PR, as we have similar cases in other places.

File: embeddings/pipeline/preprocessing_pipeline.py (hunk 13 above)
Comment (ktagowski) [reply embeddings/pipeline/preprocessing_pipeline.py]: We could discuss it, but I think I would move it to our general refactor issue

File: embeddings/transformation/hf_transformation/downsample_transformation.py (hunk 14 above)
Comment (ktagowski) [reply embeddings/transformation/hf_transformation/downsample_transformation.py:16]: I like this remark, we can have some kind of LIBRARY_SEED but I would move it to another issue and add to refactor backlog

File: tests/hps/test_hps_classification.py (hunk 15 above)
Comment (ktagowski) [reply tests/hps/test_hps_classification.py:238]: It's in tests, therefore I don't think we need additional info about that check. Test name in my opinion is sufficent and the users should now where this error occurs.

File: tests/test_callback_best_epoch.py (hunk 16 above)
Comment (ktagowski) [reply tests/test_callback_best_epoch.py:40]: Naming of this variable is problematic, as for unification we introduced one variable that stores input_column_name in pipelines. It can be single value or list/tuple. What do you think we should do with that?

File: tests/test_callback_best_epoch.py (hunk 17 above)
Comment (ktagowski) [reply tests/test_callback_best_epoch.py:81]: Added comment but in the fixture function of that object.

File: embeddings/data/dataset.py (hunk 1 above)
Comment (ktagowski) [reply embeddings/data/dataset.py:24]: > consider Base Dataset and Expandable / Extensible Dataset? Base is fine but just Dataset seems a little to general

But in this case it is General as wall as name. Currently we don't need other object

File: embeddings/pipeline/flair_preprocessing_pipeline.py [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+
+    def _get_transformation(
+        self, data_loader: FLAIR_DATALOADERS
+    ) -> FLAIR_DATASET_TRANSFORMATIONS_TYPE:
+        if isinstance(data_loader, (ConllFlairCorpusDataLoader, PickleFlairCorpusDataLoader)):
+            return DummyTransformation()
+
+        return self._get_base_transformation()
+
+    def _get_transformations(
```
Comment (djaniak) [inline embeddings/pipeline/flair_preprocessing_pipeline.py:98]: Consider using moe distinct names for these two methods

Comment (djaniak) [reply embeddings/pipeline/flair_preprocessing_pipeline.py:98]: It will be more difficult to create a bug

File: embeddings/data/datamodule.py (hunk 10 above)
Comment (AdrianSzymczak) [reply embeddings/data/datamodule.py:106]: Yes, it's standard function but nonetheless that's good to let users know at what processing is -thanks to logging libraries these can be set to output in verbose mode only for example

File: embeddings/pipeline/hf_preprocessing_pipeline.py (hunk 11 above)
Comment (AdrianSzymczak) [reply embeddings/pipeline/hf_preprocessing_pipeline.py]: How about types.py?

File: tests/test_callback_best_epoch.py (hunk 16 above)
Comment (AdrianSzymczak) [reply tests/test_callback_best_epoch.py:40]: plural names allow for list or tuple with single value inside, that's fine and singular name with list or tuple seems worse option imo - I'd go with input_column_name -> input_column_names

File: embeddings/data/dataset.py (hunk 1 above)
Comment (ktagowski) [reply embeddings/data/dataset.py]: I renamed it to `BaseDataset` and `Dataset` accordingly, we can change this concept later in the general refactor. But in my view until we add another `Dataset` class I think this naming is ok.

File: embeddings/pipeline/hf_preprocessing_pipeline.py (hunk 11 above)
Comment (ktagowski) [reply embeddings/pipeline/hf_preprocessing_pipeline.py]: I would add it to the refactor section and move all type aliases. Added to #120 ideas.

File: embeddings/pipeline/preprocessing_pipeline.py (hunk 12 above)
Comment (ktagowski) [reply embeddings/pipeline/preprocessing_pipeline.py]: Added to #120

File: embeddings/pipeline/preprocessing_pipeline.py (hunk 13 above)
Comment (ktagowski) [reply embeddings/pipeline/preprocessing_pipeline.py]: Added to #120

File: tests/test_callback_best_epoch.py (hunk 16 above)
Comment (ktagowski) [reply tests/test_callback_best_epoch.py:40]: Added to #120 This naming problem is above scope of this PR as it occurs also in other pipelines (like lighting classification). Currently name is same as in other places

File: embeddings/pipeline/flair_preprocessing_pipeline.py (hunk 18 above)
Comment (ktagowski) [reply embeddings/pipeline/flair_preprocessing_pipeline.py]: Done

**Review by AdrianSzymczak — APPROVED** [review]
